File: src/flv/video_tag.c
/**
 * video_tag.c — FLV video tag parser
 */
#include "flv/video_tag.h"
#include "core/bytes.h"
#include <string.h>
#include "librtmp2/types.h"

int lrtmp2_video_tag_parse(const uint8_t *data, size_t len, lrtmp2_video_tag_t *tag)
{
    if (!data || !tag || len < 1) return LRTMP2_ERR_INTERNAL;

    tag->frame_type = (data[0] >> 4) & 0x0F;
    tag->codec = (lrtmp2_video_codec_t)(data[0] & 0x0F);

    if (len >= 5 && (tag->codec == LRTMP2_VIDEO_H264 || tag->codec == LRTMP2_VIDEO_H265)) {
        tag->avc_packet_type = data[1];  /* 0=sequence header, 1=NALU, 2=end of sequence */
        tag->composition_time = lrtmp2_ntoh24(data + 2);
    }

    tag->data = data;
    tag->size = len;
    return LRTMP2_OK;
}

File: src/flv/video_tag.rs
//! FLV video tag parser
//!
//! Mirrors `src/flv/video_tag.h` and `src/flv/video_tag.c`.

use crate::types::{VideoTag, VideoCodec, Result, ErrorCode};

/// Parse an FLV video tag.
pub fn parse(data: &[u8], tag: &mut VideoTag) -> Result<()> {
    if data.is_empty() {
        return Err(ErrorCode::Internal);
    }

    tag.frame_type = (data[0] >> 4) & 0x0F;
    tag.codec = match data[0] & 0x0F {
        1 => VideoCodec::Jpeg,
        2 => VideoCodec::Sorenson,
        3 => VideoCodec::Screen,
        4 => VideoCodec::Vp6,
        5 => VideoCodec::Vp6a,
        6 => VideoCodec::Screen2,
        7 => VideoCodec::H264,
        12 => VideoCodec::H265,
        13 => VideoCodec::Av1,
        _ => VideoCodec::H264,
    };

    if data.len() >= 5 && (tag.codec == VideoCodec::H264 || tag.codec == VideoCodec::H265) {
        tag.avc_packet_type = data[1];
        tag.composition_time = ((data[2] as u32) << 16) | ((data[3] as u32) << 8) | (data[4] as u32);
    }

    tag.data = data.as_ptr();
    tag.size = data.len();
    Ok(())
}
